### PR TITLE
[VEN-805][VEN-806][VEN-809][VEN-845]Stable rate rebalance

### PR DESCRIPTION
**Pull request by defcon022**
## Description

This PR contains:
- Rebalancing conditions for the stable rate borrow.
- Tests for user to take borrow at both rates at a time.
- Swap rate mode with amount.

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/VToken.sol [hunk 1]
```diff
@@ -1094,11 +1098,7 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
         return actualRepayAmount;
     }
 
-    /**
-     * @dev Allows a borrower to swap his debt between stable and variable mode, or vice versa
-     * @param rateMode The rate mode that the user wants to swap to
-     **/
-    function swapBorrowRateMode(uint256 rateMode) external {
+    function swapBorrowRateModePreCalculation(address account) internal returns (uint256, uint256) {
```
Comment (chechu) [inline contracts/VToken.sol:1101]: I would add a comment including the description of the returned values

File: contracts/VToken.sol [hunk 2]
```diff
@@ -320,15 +320,19 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
         }
         uint256 utilizationRate = interestRateModel.utilizationRate(_getCashPrior(), totalBorrows, totalReserves);
-        uint256 variableBorrowRate = interestRateModel.getBorrowRate(_getCashPrior(), totalBorrows, totalReserves);
-        uint256 variableBorrows = totalBorrows - stableBorrows;
-        uint256 averageMarketBorrowRate = ((variableBorrows * variableBorrowRate) +
-            (stableBorrows * averageStableBorrowRate)) / totalBorrows;
+        uint256 averageMarketBorrowRate = _averageMarketBorrowRate();
         return
             (averageMarketBorrowRate * utilizationRate * (mantissaOne - reserveFactorMantissa)) /
             (mantissaOne * mantissaOne);
```
Comment (0xlucian) [inline contracts/VToken.sol:326]: do we need to divide by 1**36 ?

**Review by 0xlucian — APPROVED** [review]

File: contracts/VToken.sol [hunk 3]
```diff
@@ -1109,61 +1109,109 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
-        uint256 averageStableBorrowRateNew;
 
-        if (InterestRateMode(rateMode) == InterestRateMode.STABLE) {
-            require(variableDebt > 0, "vToken: swapBorrowRateMode variable debt is 0");
+        return (variableDebt, stableDebt);
+    }
 
-            stableBorrowsNew = stableBorrows + variableDebt;
-            uint256 stableBorrowRate = stableBorrowRatePerBlock();
+    function _updateStatesForStableRateSwap(
```
Comment (chechu) [inline contracts/VToken.sol:1118]: Comment needed

File: contracts/VToken.sol (hunk 2 above)
Comment (defcon022) [reply contracts/VToken.sol:326]: For balancing the equation for mantissa, we have to

File: contracts/VToken.sol [hunk 4]
```diff
@@ -1109,61 +1109,109 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
-            require(stableDebt > 0, "vToken: swapBorrowRateMode stable debt is 0");
+        accountBorrows[account].principal = variableDebt - swappedAmount;
+        accountBorrows[account].interestIndex = borrowIndex;
 
-            stableBorrowsNew = stableBorrows - stableDebt;
+        return (stableBorrowsNew, averageStableBorrowRateNew);
+    }
 
-            uint256 stableRateMantissa = accountStableBorrows[account].stableRateMantissa;
+    function _updateStatesForVariableRateSwap(
```
Comment (chechu) [inline contracts/VToken.sol:1148]: Comment needed

File: contracts/VToken.sol [hunk 5]
```diff
@@ -1109,61 +1109,109 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+        accountStableBorrows[account].interestIndex = stableBorrowIndex;
+
+        return (stableBorrowsNew, averageStableBorrowRateNew);
+    }
+
+    /**
+     * @dev Allows a borrower to swap his debt between stable and variable mode, or vice versa
+     * @param rateMode The rate mode that the user wants to swap to
+     **/
+    function swapBorrowRateMode(uint256 rateMode) external {
```
Comment (chechu) [inline contracts/VToken.sol:1185]: We could try to reuse `swapBorrowRateModeWithAmount`, calling it from this function with `type(uint256).max` (and supporting that in the called function), for example.

File: contracts/VToken.sol [hunk 6]
```diff
@@ -1468,6 +1559,43 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+        emit RebalancedStableBorrowRate(account, stableBorrowRate);
+    }
+
+    /// Validate the conditions to rebalance the stable borrow rate.
+    function validateRebalanceStableBorrowRate() public view {
+        require(rebalanceUtilizationRateThreshold > 0, "vToken: rebalanceUtilizationRateThreshold is not set.");
+        require(rebalanceRateFractionThreshold > 0, "vToken: rebalanceRateFractionThreshold is not set.");
+
+        uint256 utilizationRate = interestRateModel.utilizationRate(_getCashPrior(), totalBorrows, totalReserves);
+        uint256 variableBorrowRate = interestRateModel.getBorrowRate(_getCashPrior(), totalBorrows, totalReserves);
```
Comment (chechu) [inline contracts/VToken.sol:1589]: I think here you are calculating the variable rate given the current utilization rate, but according to the doc we should calculate the variable when the utilization rate is `rebalanceUtilizationRateThreshold`

File: contracts/VToken.sol [hunk 7]
```diff
@@ -1786,6 +1914,28 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
         emit NewAccessControlManager(oldAccessControlManager, accessControlManager);
     }
 
+    /**
+     * @notice Sets the utilization threshold for stable rate rebalancing
+     * @param utilizationRateThreshold The utilization rate threshold
+     * @custom:access Only Governance
+     */
+    function setRebalanceUtilizationRateThreshold(uint256 utilizationRateThreshold) external {
+        require(msg.sender == owner(), "only admin can set ACL address");
```
Comment (chechu) [inline contracts/VToken.sol:1923]: I would use the `AccessControlManager` to delegate these changes to VIP's

File: contracts/VToken.sol [hunk 8]
```diff
@@ -1786,6 +1914,28 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+        rebalanceUtilizationRateThreshold = utilizationRateThreshold;
+    }
+
+    /**
+     * @notice Sets the fraction threshold for stable rate rebalancing
+     * @param fractionThreshold The fraction threshold for the validation of the stable rate rebalancing
+     * @custom:access Only Governance
+     */
+    function setRebalanceRateFractionThreshold(uint256 fractionThreshold) external {
+        require(msg.sender == owner(), "only admin can set ACL address");
```
Comment (chechu) [inline contracts/VToken.sol:1934]: I would use the `AccessControlManager` to delegate these changes to VIP's

File: contracts/VToken.sol (hunk 5 above)
Comment (defcon022) [reply contracts/VToken.sol:1185]: `swapBorrowRateModeWithAmount` have addition two checks that's why I created seperate method `swapBorrowRateMode`

File: contracts/VToken.sol (hunk 3 above)
Comment (defcon022) [reply contracts/VToken.sol:1132]: Fixed, can verify at [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/9623c09d414b1d588a9416a98457f4eb8bf3dba5).

File: contracts/VToken.sol (hunk 4 above)
Comment (defcon022) [reply contracts/VToken.sol:1170]: Fixed, can verify at [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/9623c09d414b1d588a9416a98457f4eb8bf3dba5).

File: contracts/VToken.sol (hunk 7 above)
Comment (defcon022) [reply contracts/VToken.sol]: Fixed, can verify at [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/9623c09d414b1d588a9416a98457f4eb8bf3dba5).

File: contracts/VToken.sol (hunk 8 above)
Comment (defcon022) [reply contracts/VToken.sol]: Fixed, can verify at [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/9623c09d414b1d588a9416a98457f4eb8bf3dba5).

File: contracts/VTokenInterfaces.sol [hunk 9]
```diff
@@ -259,6 +265,11 @@ abstract contract VTokenInterface is VTokenStorage {
      */
     event SwapBorrowRateMode(address account, uint256 swappedBorrowMode);
 
+    /**
+     * @notice Event emitted when a borrow rate mode is swapped for account with amount
+     */
+    event SwapBorrowRateModeWithAmount(address account, uint256 swappedBorrowMode, uint256 amount);
```
Comment (chechu) [inline contracts/VTokenInterfaces.sol:271]: I think we could use the same `SwapBorrowRateMode` event, adding the swapped amount and the left amount to that event. That way we would simplify the analysis of these events

File: contracts/VToken.sol (hunk 5 above)
Comment (chechu) [reply contracts/VToken.sol:1207]: The added check is `require(variableDebt >= amount, "Insufficient amount");`, that would be satisfied if we properly support `type(uint256).max`.

Even more, I was thinking we could remove this function from the contract, delegating in the client the option to invoke `swapBorrowRateModeWithAmount` (I would rename it) with `type(uint256).max` if they want to swap 100% of the debt, as we are doing with the repay function, for example.

File: contracts/VToken.sol [hunk 10]
```diff
@@ -1468,6 +1581,46 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+    function validateRebalanceStableBorrowRate() public view {
+        require(rebalanceUtilizationRateThreshold > 0, "vToken: rebalanceUtilizationRateThreshold is not set.");
+        require(rebalanceRateFractionThreshold > 0, "vToken: rebalanceRateFractionThreshold is not set.");
+
+        uint256 utilizationRate = interestRateModel.utilizationRate(_getCashPrior(), totalBorrows, totalReserves);
+
+        /// Utilization rate is above rebalanceUtilizationRateThreshold.
+        require(utilizationRate >= rebalanceUtilizationRateThreshold, "vToken: low utilization rate for rebalacing.");
+
+        uint256 variableBorrowRate = interestRateModel.getBorrowRate(_getCashPrior(), totalBorrows, totalReserves);
```
Comment (chechu) [inline contracts/VToken.sol:1615]: We are still calculating the variable interest rate considering the current utilization rate, and we need to calculate the variable interest rate if the utilization rate is `rebalanceUtilizationRateThreshold`

File: contracts/VToken.sol [hunk 11]
```diff
@@ -1786,6 +1937,48 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+     */
+    function setRebalanceUtilizationRateThreshold(uint256 utilizationRateThreshold) external {
+        bool canCallFunction = AccessControlManager(accessControlManager).isAllowedToCall(
+            msg.sender,
+            "setRebalanceUtilizationRateThreshold(uint256)"
+        );
+        // Check caller is allowed to call this function
+        if (!canCallFunction) {
+            revert SetRebalanceUtilizationRateThresholdAdminCheck();
+        }
```
Comment (chechu) [inline contracts/VToken.sol:1953]: We could encapsulate this logic in a function, as we are doing in the new Comptroller:

```
    /// @notice Reverts if the call is not allowed by AccessControlManager
    /// @param signature Method signature
    function _checkAccessAllowed(string memory signature) internal view {
        bool isAllowedToCall = AccessControlManager(accessControl).isAllowedToCall(msg.sender, signature);

        if (!isAllowedToCall) {
            revert Unauthorized(msg.sender, address(this), signature);
        }
    }
```

I think we don't need specific errors.

File: contracts/VToken.sol (hunk 6 above)
Comment (defcon022) [reply contracts/VToken.sol]: Look at this one: https://github.com/VenusProtocol/isolated-pools/pull/157/commits/b5d8371daf6ac647a9ab0e12b9b64c24b3270eae

File: contracts/VToken.sol (hunk 10 above)
Comment (defcon022) [reply contracts/VToken.sol]: Look at this one: https://github.com/VenusProtocol/isolated-pools/commit/b5d8371daf6ac647a9ab0e12b9b64c24b3270eae

File: contracts/VToken.sol (hunk 11 above)
Comment (defcon022) [reply contracts/VToken.sol:1978]: Yes, sounds better, and we can also pass the error as the argument(If we want to use different errors).
Can we create a seperate ticket for that?

Comment (chechu) [reply contracts/VToken.sol:1978]: I would do it the same way we are doing it in IL, not using specific errors but the generic `Unauthorized`

File: contracts/VTokenInterfaces.sol (hunk 9 above)
Comment (defcon022) [reply contracts/VTokenInterfaces.sol]: Refactored, as well as for this [comment](https://github.com/VenusProtocol/isolated-pools/pull/157#discussion_r1112211425). Can verify the changes at https://github.com/VenusProtocol/isolated-pools/pull/157#discussion_r1112211425

File: contracts/BaseJumpRateModelV2.sol [hunk 12]
```diff
@@ -86,47 +86,19 @@ abstract contract BaseJumpRateModelV2 is InterestRateModel {
...
-
-        return (borrows * BASE) / (cash + borrows - reserves);
-    }
-
     /**
      * @notice Calculates the current borrow rate per block, with the error code expected by the market
-     * @param cash The amount of cash in the market
-     * @param borrows The amount of borrows in the market
-     * @param reserves The amount of reserves in the market
+     * @param utilizationRate The utilization rate as per total borrows and cash available
```
Comment (coreyar) [inline contracts/BaseJumpRateModelV2.sol:91]:
```suggestion
     * @param utilizationRate The utilization rate per total borrows and cash available
```

File: contracts/InterestRateModel.sol [hunk 13]
```diff
@@ -11,27 +11,8 @@ abstract contract InterestRateModel {
 
     /**
      * @notice Calculates the current borrow interest rate per block
-     * @param cash The total amount of cash the market has
-     * @param borrows The total amount of borrows the market has outstanding
-     * @param reserves The total amount of reserves the market has
+     * @param utilizationRate The utilization rate as per total borrows and cash available
```
Comment (coreyar) [inline contracts/InterestRateModel.sol:14]:
```suggestion
     * @param utilizationRate The utilization rate per total borrows and cash available
```

File: contracts/VTokenInterfaces.sol [hunk 14]
```diff
@@ -174,6 +174,12 @@ contract VTokenStorage {
         STABLE,
         VARIABLE
     }
+
+    /// @notice Utilization rate threshold where rebalancing condition get satisfied for stable rate borrowing.
```
Comment (coreyar) [inline contracts/VTokenInterfaces.sol:178]:
```suggestion
    /// @notice Utilization rate threshold for rebalancing stable borrowing rate.
```

File: contracts/BaseJumpRateModelV2.sol (hunk 12 above)
Comment (defcon022) [reply contracts/BaseJumpRateModelV2.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/87a380034cb29a61d3b94d7fac5d3f8f45162d43).

File: contracts/InterestRateModel.sol (hunk 13 above)
Comment (defcon022) [reply contracts/InterestRateModel.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/87a380034cb29a61d3b94d7fac5d3f8f45162d43).

File: contracts/JumpRateModelV2.sol [hunk 15]
```diff
@@ -12,17 +12,11 @@ import "./InterestRateModel.sol";
 contract JumpRateModelV2 is InterestRateModel, BaseJumpRateModelV2 {
     /**
      * @notice Calculates the current borrow rate per block
-     * @param cash The amount of cash in the market
-     * @param borrows The amount of borrows in the market
-     * @param reserves The amount of reserves in the market
+     * @param utilizationRate The utilization rate as per total borrows and cash available
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/JumpRateModelV2.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/87a380034cb29a61d3b94d7fac5d3f8f45162d43).

File: contracts/VTokenInterfaces.sol (hunk 14 above)
Comment (defcon022) [reply contracts/VTokenInterfaces.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/87a380034cb29a61d3b94d7fac5d3f8f45162d43).

File: contracts/VToken.sol [hunk 16]
```diff
@@ -1468,6 +1547,66 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+        accrueInterest();
+
+        validateRebalanceStableBorrowRate();
+        _updateUserStableBorrowBalance(account);
+
+        uint256 stableBorrowRate = stableBorrowRatePerBlock();
+
+        accountStableBorrows[account].stableRateMantissa = stableBorrowRate;
+
+        emit RebalancedStableBorrowRate(account, stableBorrowRate);
```
Comment (chechu) [inline contracts/VToken.sol:1588]: We could add the previous stable rate and the new one

File: contracts/VToken.sol (hunk 11 above)
Comment (chechu) [reply contracts/VToken.sol:1939]: VEN-1123

File: contracts/VToken.sol [hunk 17]
```diff
@@ -1468,6 +1547,66 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
...
+        require(rebalanceRateFractionThreshold > 0, "vToken: rebalanceRateFractionThreshold is not set.");
+
+        uint256 utilRate = utilizationRate(_getCashPrior(), totalBorrows, totalReserves);
+
+        /// Utilization rate is above rebalanceUtilizationRateThreshold.
+        require(utilRate >= rebalanceUtilizationRateThreshold, "vToken: low utilization rate for rebalacing.");
+
+        uint256 variableBorrowRate = interestRateModel.getBorrowRate(rebalanceUtilizationRateThreshold);
+
+        /// Average market borrow rate should be less than the rebalanceRateFractionThreshold fraction of variable borrow rate.
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/VToken.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/f00e9d9196bd7c1c102aa9991ecdc0253411e92f).

File: contracts/VTokenInterfaces.sol [hunk 18]
```diff
@@ -255,9 +261,9 @@ abstract contract VTokenInterface is VTokenStorage {
     );
 
     /**
-     * @notice Event emitted when a borrow rate mode is swapped for account
+     * @notice Event emitted when a borrow rate mode is swapped for account with amount
      */
-    event SwapBorrowRateMode(address account, uint256 swappedBorrowMode);
+    event SwapBorrowRateMode(address account, uint256 swappedBorrowMode, uint256 amount);
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/VTokenInterfaces.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/f00e9d9196bd7c1c102aa9991ecdc0253411e92f).

File: contracts/VToken.sol (hunk 16 above)
Comment (defcon022) [reply contracts/VToken.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/f00e9d9196bd7c1c102aa9991ecdc0253411e92f).

File: contracts/VTokenInterfaces.sol [hunk 19]
```diff
@@ -319,6 +325,21 @@ abstract contract VTokenInterface is VTokenStorage {
      */
     event UpdatedUserStableBorrowBalance(address borrower, uint256 updatedPrincipal);
 
+    /**
+     * @notice Event emitted on stable rate rebalacing
+     */
+    event RebalancedStableBorrowRate(address account, uint256 stableRateMantissa);
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/VTokenInterfaces.sol]: Fixed, [commit](https://github.com/VenusProtocol/isolated-pools/pull/157/commits/f00e9d9196bd7c1c102aa9991ecdc0253411e92f).

**Review by chechu — APPROVED** [review]
LGTM, good job!

**Pull request merged**